src/components/WebCLI/WebCLI.tsx

```diff
@@ -127,22 +127,53 @@ const WebCLI = ({
     const conn = new Connection({
       address: wsAddress,
       onopen: () => {
-        // Unused handler.
+        // Close this websocket connection if it was in CONNECTING state
+        // while a new websocket connection was established.
+        if (conn.websocket !== connection.current?.websocket) {
+          conn.disconnect();
+        }
       },
       onclose: () => {
         // Unused handler.
       },
+      onerror: (error) => {
+        setInlineError(
+          InlineErrors.CONNECTION,
+          typeof error === "string" ? error : "Unknown error.",
+        );
+      },
       messageCallback: (message: string) => {
         wsMessageStore.current = wsMessageStore.current + message;
         setOutput(wsMessageStore.current);
       },
     }).connect();
-    setConnection(conn);
-    return () => {
-      conn.disconnect();
-    };
+    if (connection.current?.isOpen()) {
+      connection.current?.disconnect();
+    }
+    connection.current = conn;
   }, [setInlineError, wsAddress]);
 
+  useEffect(
+    () => () => {
+      // Cleanup gets triggered twice in development environment due to
+      // React's StrictMode. As it gets triggered in rapid successions, it
+      // closes the previous WebSocket connection while it is in CONNECTING
+      // state. This triggers an error in the closed WebSocket connection, which
+      // subsequently displays an error message in the WebCLI. To fix this
+      // issue, in development environment we only close the WebSocket
+      // connection if it is in OPEN state. In production environment, as the
+      // cleanup runs only once, we just close the last connection, even if it
+      // is in OPEN state.
+      if (
+        process.env.NODE_ENV !== "development" ||
+        connection.current?.isOpen()
+      ) {
+        connection.current?.disconnect();
+      }
```

Comment:
Close the connection even if it is pending on unmount (this will also close the first websocket in strict dev mode, but that's OK as a new one will get created).

```suggestion
      if (connection.current?.isActive()) {
        connection.current?.disconnect();
      }
```

Reply:
I moved the check for `isActive` inside the `disconnect` method of `Connection` class. I was thinking that this approach would be less error prone.